### DISCUSSION TITLE
Is LocalSplitter included by StaticSingleAssignmentFormer？

### DISCUSSION THREAD

**Discussion by Liyw979** [soot-oss/SootUp#1122] [Q&A] [answered]
https://soot-oss.github.io/SootUp/latest/bodyinterceptors/#staticsingleassignmentformer
Do we need to enable  LocalSplitter before using StaticSingleAssignmentFormer?

**Comment by swissiety** [comment]
Having the Localsplitter enabled is always a good idea to create better (valid) Jimple. Maybe @wangzun66 has some insights about this as she is looking into improving/fixing SSA, currently.

**Comment by wangzun66** [comment, marked as answer]
In fact, LocalSplitter and StaticSingleAssignmentFormer are two different BodyInterceptors. Their relationship is not one of containment.  When a CFG has no fork-joins, as shown in [LocalSplitter Example_1](https://github.com/soot-oss/SootUp/blob/develop/docs/assets/figures/LocalSplitter%20Example_1.png), their effects are similar, although the renaming way differs slightly. When a CFG contains fork-joins,  as shown in [LocalSplitter Example_2](https://github.com/soot-oss/SootUp/blob/develop/docs/assets/figures/LocalSplitter%20Example_2.png), their effects are very different. LocalSplitter treats the same local variables that appears in both branches as a single local variable. In contrast, StaticSingleAssignmentFormer treats them as two different local variables and adds an assignment with a phi-expression at their merge point to distinguish them.
